ci-operator/step-registry/nodes/readiness/nodes-readiness-commands.sh

```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+if [ -f "${SHARED_DIR}/proxy-conf.sh" ] ; then
+  source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+if ! command -v yq-v4 &> /dev/null; then
+  echo "[WARN] yq-v4 is not installed. Downloading it..."
+  mkdir -p /tmp/bin
+  this_pod_arch=$(uname -m | sed 's/x86_64/amd64/;s/aarch64/arm64/')
+  curl -L "https://github.com/mikefarah/yq/releases/download/v4.30.5/yq_linux_${this_pod_arch}" -o /tmp/bin/yq-v4
+  chmod +x /tmp/bin/yq-v4
+  export PATH=${PATH}:/tmp/bin
+fi
+
+function debug() {
+  echo "[DEBUG] Current machinesets, machines and nodes are:"
+  set +e
+  for resource in machinesets.machine.openshift.io machines.machine.openshift.io nodes; do
+    oc -n openshift-machine-api get "${resource}" -owide | tee "${ARTIFACT_DIR}/${resource}.txt"
+    oc -n openshift-machine-api get "${resource}" -oyaml | tee "${ARTIFACT_DIR}/${resource}.yaml"
+    oc -n openshift-machine-api describe "${resource}"   | tee "${ARTIFACT_DIR}/${resource}-describe.txt"
+  done
+  set -e
+}
+
+# get_ready_nodes_count returns the number of ready nodes
+function get_ready_nodes_count() {
+  oc get nodes \
+    -o jsonpath='{range .items[*]}{.metadata.name}{","}{.status.conditions[?(@.type=="Ready")].status}{"\n"}{end}' | \
+    grep -c -E ",True$"
+}
+
+# wait_for_nodes_readiness loops until the number of ready nodes objects is equal to the desired one
+# It takes 3 arguments:
+# - expected_nodes: the number of nodes that should be ready
+# - max_retries: the maximum number of retries before failing. Default: 5.
+# - period: the time to wait between retries in minutes. Default: 1.
+function wait_for_nodes_readiness()
+{
+  local expected_nodes=${1}
+  local max_retries=${2:-20}
+  local period=${3:-1}
+  echo "[INFO] Waiting for ${expected_nodes} nodes to be ready..."
+  for i in $(seq 1 "${max_retries}") max; do
+    if [ "${i}" == "max" ]; then
+      echo "[ERROR] Timeout reached. ${expected_nodes} ready nodes expected, found ${ready_nodes}... Failing."
+      debug
+      return 1
+    fi
+    ready_nodes=$(get_ready_nodes_count)
+    if [ "${ready_nodes}" == "${expected_nodes}" ]; then
+        echo "[INFO] Found ${ready_nodes}/${expected_nodes} ready nodes, continuing..."
+        return 0
+    fi
+    echo "[INFO] - ${expected_nodes} ready nodes expected, found ${ready_nodes}..." \
+      "Waiting ${period}min before retrying (timeout in $(( (max_retries - i + 1) * (period) ))min)..."
+    sleep "${period}m"
+  done
+}
+
+expected_nodes=$(( $(yq-v4 -r '.controlPlane.replicas // '"${CONTROL_PLANE_REPLICAS:-3}" "${SHARED_DIR}/install-config.yaml") ))
+expected_nodes=$(( expected_nodes + $(yq-v4 -r '.compute[0].replicas // '"${COMPUTE_NODE_REPLICAS:-3}" "${SHARED_DIR}/install-config.yaml")))
+
+wait_for_nodes_readiness "${expected_nodes}"
```

Comment:
For the nodes provisioned in edge zone, they will take a longer time to get ready, I suggest to max_retries for them, like:
```
if edge-node; then
  wait_for_nodes_readiness "${expected_nodes}"
else
  wait_for_nodes_readiness "${expected_nodes}" 30
fi
```

Reply:
I updated this for both cases to a 30-minute timeout. I still like the idea of keeping these steps as simple as possible, and thanks to Jianlin's previous comments, we've already reduced the surface area of conditional branching.

This might mean that we use 10 minutes more in non-edge cases before failing. However, this step's target is mostly to 'sleep' until all nodes are ready, not to catch failures. 

In the majority of the cases, I expect a failure to occur at the earlier installation steps and the nodes readiness step to conclude successfully in a few seconds/minutes: therefore, without ever reaching the timeout but still allowing any further validation steps from needing to implement their own way to verify nodes readiness before doing another specific check. (This also seems to be the reason why the installer doesn't check all-nodes readiness in its code, but for the CI, I feel safer knowing the nodes are ready **before** running further steps).

That's also why I would really like (a) the node-readiness step to be set for all the consumers of the ipi-install chain, (b) to roll out the validation of the nodes architecture while we roll out the support for mix arch day0 control/compute in the installer (i.e., while the CONTROL_ARCH and COMPUTE_ARCH vars are rolled out in the CI).